src/v/transform/txn_reader.cc

```diff
@@ -0,0 +1,219 @@
+        process(std::move(batch));
+        return ss::make_ready_future<ss::stop_iteration>(
+          ss::stop_iteration::no);
+    }
+
+    drain_result end_of_stream() { return std::move(_result); }
+
+private:
+    void process(model::record_batch batch) {
+        model::record_batch_attributes attrs = batch.header().attrs;
+        // It's possible that there are only control batches, which then we
+        // don't have to filter out any aborts.
+        if (attrs.is_transactional()) {
+            if (!_result.txn_range) {
+                _result.txn_range.emplace(
+                  batch.base_offset(), batch.last_offset());
+            }
+            _result.txn_range->second = batch.last_offset();
+        }
+        _result.batches.push_back(std::move(batch));
+    }
+
+    drain_result _result;
+};
+
+bool is_aborted(
+  const model::record_batch& batch,
+  const absl::btree_set<model::tx_range, std::greater<>>& aborted_txns) {
+    model::producer_identity pid = {
+      batch.header().producer_id, batch.header().producer_epoch};
+    auto it = aborted_txns.lower_bound(
+      {pid, batch.base_offset(), model::offset::max()});
+
+    if (it == aborted_txns.end()) {
+        return false;
+    }
+    // If the pids match, and the batch lies within the abort range, then this
+    // batch is aborted.
+    return it->pid == pid && batch.base_offset() >= it->first
+           && batch.last_offset() <= it->last;
+}
+
+std::unique_ptr<model::record_batch_reader::impl> make_txn_filtered_reader(
+  ss::chunked_fifo<model::record_batch> batches,
+  const absl::btree_set<model::tx_range, std::greater<>>& aborted_txns) {
+    ss::chunked_fifo<model::record_batch> filtered;
+    for (auto& batch : batches) {
+        auto attrs = batch.header().attrs;
+        if (attrs.is_transactional() && is_aborted(batch, aborted_txns)) {
+            continue;
+        }
+        if (attrs.is_control()) {
+            continue;
+        }
+        filtered.push_back(std::move(batch));
+    }
+    return model::make_fragmented_memory_record_batch_reader(
+             std::move(filtered))
+      .release();
+}
+
+} // namespace
+
+read_committed_reader::read_committed_reader(
+  std::unique_ptr<aborted_transaction_tracker> tracker,
+  model::record_batch_reader reader)
+  : _tracker(std::move(tracker))
+  , _underlying(std::move(reader).release()) {}
+
+void read_committed_reader::print(std::ostream& os) {
+    os << "transform::txn_reader{}";
+}
+
+bool read_committed_reader::is_end_of_stream() const {
+    return _underlying->is_end_of_stream();
+}
+
+ss::future<model::record_batch_reader::storage_t>
+read_committed_reader::do_load_slice(
+  model::timeout_clock::time_point deadline) {
+    model::record_batch_reader::storage_t loaded
+      = co_await _underlying->do_load_slice(deadline);
+    // We delete the tracker when we've filtered the stream already.
+    if (!_tracker || !contains_control_or_txn_batch(loaded)) {
```

Comment:
q: this check means that this reader can only load one slice, right? That seems a bit counter intuitive given how batch readers are used elsewhere, they usually load new slice in each load_slice() invocation.

Reply:
> q: this check means that this reader can only load one slice, right? That seems a bit counter intuitive given how batch readers are used elsewhere, they usually load new slice in each load_slice() invocation (or empty if eos)

I'm confused what you mean by here, the reader can load multiple slices? This is effectively the decorator pattern: https://refactoring.guru/design-patterns/decorator

We have tests that generator multiple slices too.

Reply:
> This reader is optimistic supporting a passthrough reading style until a
transaction is found, at which point we buffer all records into memory
filter out aborted/control records before continuing.

This `if` statement is basically:

If we have already buffered and filtered out records then just pass the loaded slice up to the next layer OR if the batch doesn't have a control or txn batch, then pass it up to the next layer (no processing is needed)

Reply:
I should get my eyesight checked 🙈